### DISCUSSION TITLE
Why is the `CutSet.cut_into_windows()` using Lazy?

### DISCUSSION THREAD

**Discussion by ArthLeu** [lhotse-speech/lhotse#1211] [Q&A]
Is this necessary to return a Lazy CutSet instead of an eager one? My use case is cutting a cutset into windows and then using `PrecomputedFeatures()` to load features. But `PrecomputedFeatures()` cannot take lazy cutset, so I had to convert the set to eager again, which caused computation overhead.

So essentially this is like: eager -> `cut_into_windows()` -> lazy -> `to_eager()` -> eager -> `PrecomputedFeatures()`

Just timed it, `to_eager()` almost takes half of the time in each data loading round.

Is there a better way to do this?

**Comment by pzelasko** [comment]
> Just timed it, to_eager() almost takes half of the time in each data loading round. 

Had we done the cutting into windows eagerly, you'd just pay the same price but at the time of calling `cut_into_windows()`. I think it might be possible to come up with a cutting algorithm with lesser computational complexity for it which would speed things up (IIRC the majority of the time is being spent on checking which supervisions should be kept in a given window, but we're already using interval tree indexes to make it more efficient).

Can you share a bit more about what are you trying to do? I'm not sure what to advise you without more context.

**Reply by ArthLeu** [reply]
> Can you share a bit more about what are you trying to do? I'm not sure what to advise you without more context.

Sure. I have some Lhotse CutSet with precomputed Kaldi features enclosed. Now the previous workflow with my VAD model using Kaldi starts first with cutting them into fix length chunks (or windows), usually a few seconds. There wasn't any bottleneck issue before incorporating Lhotse. When I started using [this VAD recipe](https://github.com/lhotse-speech/lhotse/blob/master/lhotse/dataset/vad.py) together with `cut_into_windows`, there was a significant slowdown in data part. And then I tracked down to the `to_eager()`, which looks to be mandatory after `cut_into_windows()` and before `PrecomputedFeatures`.

**Reply by pzelasko** [reply]
Are you using some sampler class (e.g. `DynamicCutSampler`) to create mini-batches? In general, Dataset classes and `PrecomputedFeatures` are intended to be used with mini-batches rather than full CutSets.

As a side note I can't see where `PrecomputedFeatures` would fail, do you have any error message?

**Reply by ArthLeu** [reply]
Yes, I've been using SimpleCutSampler and now testing the DynamicCutSampler. 

The error message when calling `PrecomputedFeatures` without `to_eager()` is
```
"LazyFlattener does not support __len__ because it would require "
            "iterating over the whole iterator, which is not possible in a lazy fashion. "
            "If you really need to know the length, convert to eager mode first using "
            "`.to_eager()`. Note that this will require loading the whole iterator into memory."
```
where LazyFlattener is the return type of `cut_into_windows()`.
